Data/Primitive/Array.hs

```diff
@@ -586,18 +586,28 @@ mapArray' f a =
 
 -- | Create an array from a list of a known length. If the length
 -- of the list does not match the given length, this throws an exception.
+
+-- Note [fromListN]
+-- ~~~~~~~~~~~~~~~~
+-- We want arrayFromListN to be a "good consumer" in list fusion, so we define
+-- the function using foldr and inline it to help fire fusion rules.
+-- If fusion occurs with a "good producer", it may reduce to a fold on some
+-- structure. In certain cases (such as for Data.Set) GHC is not be able to
+-- optimize the index to an unboxed Int# (see GHC #24628), so we explicitly use
+-- an Int# here.
 arrayFromListN :: Int -> [a] -> Array a
+{-# INLINE arrayFromListN #-}
 arrayFromListN n l =
   createArray n (die "fromListN" "uninitialized element") $ \sma ->
-    let go !ix [] = if ix == n
+    let z ix# = if I# ix# == n
           then return ()
           else die "fromListN" "list length less than specified size"
-        go !ix (x : xs) = if ix < n
+        f x k = GHC.Exts.oneShot $ \ix# -> if I# ix# < n
```

Comment:
@andrewthad You didn't need to drop support for ghc-8.0 in #426, since `oneShot` is available form `ghc-prim` for a very long time, so it could have been imported from `GHC.Magic` for all GHC versions, instead of relying on CPP. Especially since the test suite already depends on `ghc-prim`. Seems like a very minor thing to drop support for a whole ghc version, but I personally not gonna cry over it :smiling_face_with_tear:

Reply:
Yes, but I would've needed to add the dependency on ghc-prim back to primitive, and it was so nice to finally have it gone. Dropping GHC 8.0 has the additional advantage of letting me delete several shims in https://github.com/haskell/primitive/pull/427. I'll probably end up dropping support for all GHCs earlier than 8.6 soon.